debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/OracleConnectorTask.java

```diff
@@ -139,4 +164,100 @@ public void doStop() {
+            return false;
+        }
+        if (!schema.historyExists()) {
+            LOGGER.warn("Database history was not found but was expected");
+            if (config.getSnapshotMode().shouldSnapshotOnSchemaError()) {
+                // But check to see if the server still has those redo log coordinates ...
+                if (!isRedoLogAvailable(config, offset)) {
+                    throw new DebeziumException("The connector is trying to read redo log starting at " + offset.getSourceInfo() + ", but this is no longer "
+                            + "available on the server. Reconfigure the connector to use a snapshot when needed.");
+                }
+                LOGGER.info("The db-history topic is missing but we are in {} snapshot mode. " +
+                        "Attempting to snapshot the current schema and then begin reading the redo log from the last recorded offset.",
+                        OracleConnectorConfig.SnapshotMode.SCHEMA_ONLY_RECOVERY);
+            }
+            else {
+                throw new DebeziumException("The db history topic is missing. You may attempt to recover it by reconfiguring the connector to "
+                        + OracleConnectorConfig.SnapshotMode.SCHEMA_ONLY_RECOVERY);
+            }
+            schema.initializeStorage();
+            return true;
+        }
+        schema.recover(partition, offset);
+        return false;
+    }
+
+    /**
+     * Determine whether the redo log position as set on the {@link OracleOffsetContext} is available in the server.
+     *
+     * @return {@code true} if the server has the redo log coordinates, or {@code false} otherwise
+     */
+    protected boolean isRedoLogAvailable(OracleConnectorConfig config, OracleOffsetContext offset) {
+        String lcrStr = offset.getLcrPosition();
+        Scn scn;
+
+        if (lcrStr != null) {
+            scn = LcrPosition.valueOf(lcrStr).getScn();
+        }
+        else {
+            scn = offset.getScn();
+        }
+
+        if (scn == null || scn.isNull()) {
+            return false;
+        }
+
+        try {
+            Scn oldestScn = LogMinerStreamingChangeEventSource.getFirstScnInLogs(jdbcConnection, config.getLogMiningArchiveLogRetention(),
+                    config.getLogMiningArchiveDestinationName());
+            Scn currentScn = jdbcConnection.getCurrentScn();
+            // And compare with the one we're supposed to use.
+            if (scn.compareTo(oldestScn) < 0 || scn.compareTo(currentScn) > 0) {
+                LOGGER.info("Connector requires redo log scn '{}', but Oracle only has {}-{}", scn, oldestScn, currentScn);
+                return false;
+            }
+            else {
+                LOGGER.info("Oracle has the redo log scn '{}' required by the connector", scn);
+                return true;
+            }
+        }
+        catch (SQLException e) {
+            throw new DebeziumException(e);
+        }
+    }
```

Comment:
As mentioned above, this logic should be split between its Xstream and LogMiner parts and moved to the appropriate `StreamingAdapter` implementation class.  The adapter can be obtained by calling `OracleConnectorConfig#getAdapter`.

Reply:
Since I don't know much about xstream's API, I deleted the code. And LogMinerStreamingChangeEventSource already has some logic for scn verification

Reply:
Thinking about this code a bit more, I think what you did is probably the right call.  

To give some background, Oracle starts a server-side process when using Xstream and the Outbound Server process is responsible for picking up changes and putting them into a queue that the Xstream client will pull from when we connect.  So with Xstream we don't really need to be concerned with whether the SCN is available in the redo logs.

With LogMiner this is entirely different since the ownership of getting change events is entirely on the connector.  We need to verify that the offset SCN is still available and throw an exception if it isn't.  

So with that, leaving the code as-is for this check and removing this method entirely is the right path since really only the LogMiner implementation would face problems if the offset wasn't in the redo logs.